Added button to select attributes from largest geometry when merging features

uclaros opened the pull request:
## Description
Inspired by #42194, I didn't want to make this automatic since there is some logic for fields with default values present.
![image](https://user-images.githubusercontent.com/11358178/110856306-26ba3000-82c0-11eb-86d1-396a33a85148.png)
Only enabled for Polygon and Line layers.

Also changed the *autoDefault* property for the first three buttons so that they do not capture the *enter* key and allow it to accept the dialog.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

uclaros committed:
Added button to select attributes from largest geometry when merging features

nyalldawson commented on src/app/qgsmergeattributesdialog.cpp:
```diff
@@ -460,21 +484,49 @@ void QgsMergeAttributesDialog::mFromSelectedPushButton_clicked()
+      for ( ; f_it != mFeatureList.constEnd(); ++f_it )
+      {
+        double featureLength = f_it->geometry().length();
+        if ( featureLength > maxValue )
```
if all lengths are zero then featureId will never be set. Possibly a dumb corner case, but maybe this should be >= or maxValue should start at -1 ?

uclaros replied:
This is kind of on purpose. There is no reason to select one featureId randomly when there is not one larger geometry, and there is line 528 to safeguard uninitialized use of the featureId.

nyalldawson replied:
Ah - gotcha. All good then!

nyalldawson commented on src/ui/qgsmergeattributesdialogbase.ui:
```diff
@@ -57,7 +60,53 @@
+     <item>
+      <widget class="QLabel" name="mTakeLargestAttributesLabel">
+       <property name="text">
+        <string>Take attributes from largest geometry</string>
```
```suggestion
        <string>Take attributes from feature with the largest geometry</string>
```

nyalldawson commented:
Nice feature!

Regarding the icon -- I'd suggest either:
1 . A shaded with polygon with a ">" symbol overlaid
2. Having two polygons, one clearly larger than the other and having the larger polygon appear in the yellow selected color

uclaros commented:
> Regarding the icon -- I'd suggest either:
> 1 . A shaded polygon with a ">" symbol overlaid
> 2. Having two polygons, one clearly larger than the other and having the larger polygon appear in the yellow selected color

I'll see what I can come up with!

Maybe I could also support Multi-Points too and take the geometry with the most parts?

uclaros committed:
Fix label wording

nyalldawson commented:
> I'll see what I can come up with!

If you're not feeling inspired then just reuse an existing icon instead of adding a new one (pick any old random one :P ) and open a ticket requesting a new icon, and I'll add it to https://github.com/qgis/QGIS/projects/1

uclaros committed:
Updated icon

uclaros committed:
Also support MultiPoints

uclaros committed:
Also select row with largest feature so it gets highlighted

uclaros commented:
I updated the icon and added support for selecting multipoints based on partCount.
![image](https://user-images.githubusercontent.com/11358178/110931739-48a0ca80-8333-11eb-99d0-536420c15e24.png)

The appropriate row is now also selected so it gets highlighted for visual confirmation.
